### PR TITLE
[ENH] Pivot longer faster

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Primary goal was to make it faster, and make it faster than pd.wide_to_long if possible
- Cleaner syntax
- dtypes change no longer needed within the `pivot_longer` function, it can be applied outside with `astype`

**This PR improves #755 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [x] Build the docs locally.
- [x] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by samukweku** [comment]
@loganthomas ; need your help here with the `darglint` errors. Not sure what to do.

**Comment by loganthomas** [comment]
@samukweku @ericmjl 
Looks like a recent release of `darglint` is trying to handle nested function definitions (see the `darglint` changelog for `1.5.6` [here](https://github.com/terrencepreilly/darglint/blob/master/CHANGELOG.md)).

I did some investigation and the pipeline is failing at these locations (`chemistry` line 130, `functions` line 280, 1623 , and 4482):
Screen Shot 2020-11-22 at 09 59 46

If you take a look at the source code, we are using `lambda` functions with the arguments `x` and `d`. This is what is causing the error.

We have two options:

1. Roll back darglint to a version that doesn't try to parse nested functions (we were using `1.5.4` when we first introduced `darglint`)

2. Append a `darglint` ignore statement (`.. # noqa: DAR101 <var>`)  as the last line of the docstrings that use a lambda function within the function's source code body.
  - __Note__: the `..` prefix is important as it will ensure the `noqa` comment doesn't appear in the docs
 - __Note__: the `var` needs to be filled in with whatever input variable the `lambda` function takes. So, for `groupby_topk` that would be `.. # noqa: DAR101 d`. For the others, it would be `.. # noqa: DAR101 x`
- Example

```python
@pf.register_dataframe_method
def groupby_topk(
    df: pd.DataFrame,
    groupby_column_name: Hashable,
    sort_column_name: Hashable,
    k: int,
    sort_values_kwargs: Dict = None,
) -> pd.DataFrame:
    """
    
    ...

    :param df: A pandas dataframe.
    :param groupby_column_name: Column name to group input dataframe `df` by.
    :param sort_column_name: Name of the column to sort along the
        input dataframe `df`.
    :param k: Number of top rows to return from each group after sorting.
    :param sort_values_kwargs: Arguments to be passed to sort_values function.
    :returns: A pandas dataframe with top `k` rows that are grouped by
        `groupby_column_name` column with each group sorted along the
        column `sort_column_name`.
    :raises ValueError: if `k` is less than 1.
    :raises ValueError: if `groupby_column_name` not in dataframe `df`.
    :raises ValueError: if `sort_column_name` not in dataframe `df`.
    :raises KeyError: if `inplace:True` is present in `sort_values_kwargs`.

    .. # noqa: DAR101 d
    """  # noqa: E501
```

I tested out both options and they both worked on my local machine.

@ericmjl any preference here? I filled out an issue on `darglints` GitHub page [here](https://github.com/terrencepreilly/darglint/issues/142).

**Comment by samukweku** [comment]
> Roll back darglint to a version that doesn't try to parse nested functions (we were using 1.5.4 when we first introduced darglint)

@loganthomas @ericmjl I believe the first option is the better option, less intrusive changes within the code

**Comment by loganthomas** [comment]
@samukweku @ericmjl 
Looks like `darglint` already released a new version to help us out 🙌 . 

See `1.5.7` [here](https://github.com/terrencepreilly/darglint/blob/master/CHANGELOG.md)

Let's see if we can re-run the pipeline without any edits/rollbacks and get it to pass.

**Comment by loganthomas** [comment]
Success! 🎉  
Now the pipeline is only failing due to a long line:

```
./janitor/functions.py:4890:80: E501 line too long (88 > 79 characters)
```

**Comment by samukweku** [comment]
@loganthomas Thanks! I will work on it once I get back to my computer.

**Comment by ericmjl** [comment]
Wonderful, thank you @loganthomas and @samukweku!

**Comment by samukweku** [comment]
@ericmjl @loganthomas @hectormz  and others, as a general question, is it worthwhile to add speed tests, comparing to say pandas wide to long ?

**Comment by ericmjl** [comment]
> @ericmjl @loganthomas @hectormz and others, as a general question, is it worthwhile to add speed tests, comparing to say pandas wide to long?

Hey @samukweku! We haven't, though I do know @zbarry has had concerns over the speed of some functions before. (Prior context: the concerns surrounded whether we should do an explicit `df.copy()` on each function call, but... I think we generally worked around that problem by relying on calling pandas'/other package's native API.)

Adding speed tests is a tricky matter though. What were you thinking of doing with speed tests? Was it to sanity check that the function fully executes within an upper-bounded time budget? (Anything fancier, like checking for speed regressions, need fancier setups I think.)

File: tests/functions/test_pivot_longer.py [hunk 1]
```diff
@@ -76,6 +76,472 @@ def test_df():
     )
 
 
+@pytest.fixture
+def df_multi():
```
Comment (ericmjl) [inline tests/functions/test_pivot_longer.py:80]:
```suggestion
def df_multi():
    """MultiIndex dataframe fixture."""
```

File: tests/functions/test_pivot_longer.py [hunk 2]
```diff
@@ -76,6 +76,472 @@ def test_df():
...
+]
+names_pattern_type_check = [
+    (df_checks, "rar", 1),
+    (df_checks, {1: "rar"}, {"\\d+"}),
+]
+
+
+@pytest.mark.parametrize("df,index", index_type_checks)
+def test_type_index(df, index):
+    "Raise TypeError if wrong type is provided for the `index`."
```
Comment (ericmjl) [inline tests/functions/test_pivot_longer.py]: I think you intend for this line to be a docstring. The same issue occurs in other places. There's a lot of them, so I won't mark them individually but do see if you can fix them up. (If need be, I can hack out a snippet that isolates the functions with non-docstring docstrings, which could help.)

File: tests/functions/test_pivot_longer.py [hunk 3]
```diff
@@ -76,6 +76,472 @@ def test_df():
...
+    and wrong type is supplied for `names_to`.
+    """
+    with pytest.raises(TypeError):
+        df_checks.pivot_longer(
+            names_to={"variable, value"}, names_pattern=["1", "rar"]
+        )
+
+
+def test_column_level_wrong_type(df_multi):
+    "Raise TypeError if wrong type is provided for `column_level`."
```
Comment (ericmjl) [inline tests/functions/test_pivot_longer.py:232]: docstring

File: tests/functions/test_pivot_longer.py [hunk 4]
```diff
@@ -76,6 +76,472 @@ def test_df():
...
+    with pytest.raises(TypeError):
+        df_multi.pivot_longer(index="name", column_level={0})
+
+
+def test_sort_by_appearance(test_df):
+    "Raise error if `sort_by_appearance` is not boolean."
```
Comment (ericmjl) [inline tests/functions/test_pivot_longer.py:238]: docstring

File: tests/functions/test_pivot_longer.py [hunk 5]
```diff
@@ -76,6 +76,472 @@ def test_df():
...
+def test_ignore_index(test_df):
+    "Raise error if `ignore_index` is not boolean."
+    with pytest.raises(TypeError):
+        test_df.pivot_longer(
+            names_to=[".value", "value"], names_sep="_", ignore_index="TRUE"
+        )
+
+
+def test_duplicate_dot_value(test_df):
+    "Raise error if `names_to` contains more than 1 `.value."
```
Comment (ericmjl) [inline tests/functions/test_pivot_longer.py]: docstring.

Also, do you mean: "Raise error if `names_to` contains more than one value"?

**Review by ericmjl — CHANGES_REQUESTED** [review]
Cool stuff! Thanks, @samukweku, appreciate the work you've been doing to carry the project forward this while. I just surfaced up, Thanksgiving has given me a bit of breathing room to do PR reviews again :smile:.

**Comment by samukweku** [comment]
@ericmjl , I was thinking more of `pivot_longer can do the same thing as pandas' wide_to_long, and in some cases, can be faster/slower`. So, using jupyter's `%timeit` to show the timings.

**Comment by ericmjl** [comment]
@samukweku looks like this one only failed on Flake8. Once it's fixed, we can merge.

**Comment by samukweku** [comment]
@ericmjl errors fixed. I'll wait till weekend for another review before merging, if that's ok.

**Review by ericmjl — APPROVED** [review]

**Comment by ericmjl** [comment]
Thanks @samukweku! I'm approving, and going to merge :)

**Pull request merged**